src/stock.rs

```diff
@@ -0,0 +1,423 @@
+    /// name.
+    ///
+    /// If `from_id` is not `0`, any trailing dot is removed from the
+    /// first stock string created so far.  If the `from_id` contact is
+    /// the user itself, i.e. `DC_CONTACT_ID_SELF` the string is used
+    /// itself as param to the [StockMessage::MsgActionByMe] stock string
+    /// resulting in a string like "Member Alice added by me." (for
+    /// [StockMessage::MsgAddMember] as `id`).  If the `from_id` contact
+    /// is any other user than the contact's display name is looked up and
+    /// used as the second parameter to [StockMessage::MsgActionByUser] with
+    /// again the original stock string being used as the first parameter,
+    /// resulting in a string like "Member Alice added by Bob.".
+    pub fn stock_system_msg(
+        &self,
+        id: StockMessage,
+        param1: impl AsRef<str>,
+        param2: impl AsRef<str>,
+        from_id: u32,
+    ) -> String {
+        let insert1 = if id == StockMessage::MsgAddMember || id == StockMessage::MsgDelMember {
+            unsafe {
+                let param1_c = CString::new(param1.as_ref()).unwrap();
+                let contact_id = dc_lookup_contact_id_by_addr(self, param1_c.as_ptr());
+                if contact_id != 0 {
+                    let contact = dc_get_contact(self, contact_id);
+                    let displayname = dc_contact_get_name_n_addr(contact);
+                    let ret = to_string(displayname);
+                    free(contact as *mut libc::c_void);
+                    free(displayname as *mut libc::c_void);
+                    ret
+                } else {
+                    param1.as_ref().to_string()
+                }
+            }
+        } else {
+            param1.as_ref().to_string()
+        };
+        let action = self.stock_string_repl_str2(id, insert1, param2.as_ref().to_string());
+        let action1 = action.trim_end_matches('.');
+        match from_id {
+            0 => action,
+            1 => self.stock_string_repl_str(StockMessage::MsgActionByMe, action1), // DC_CONTACT_ID_SELF
+            _ => unsafe {
+                let contact = dc_get_contact(self, from_id);
+                let displayname = dc_contact_get_display_name(contact);
+                let ret = self.stock_string_repl_str2(
+                    StockMessage::MsgActionByUser,
+                    action1,
+                    as_str(displayname),
+                );
+                free(contact as *mut libc::c_void);
+                free(displayname as *mut libc::c_void);
+                ret
+            },
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::*;
+
+    use std::ffi::CString;
+
+    use crate::constants::DC_CONTACT_ID_SELF;
+    use crate::context::dc_context_new;
+    use crate::types::uintptr_t;
+
+    use num_traits::ToPrimitive;
+
+    #[test]
+    fn test_enum_mapping() {
+        assert_eq!(StockMessage::NoMessages.to_usize().unwrap(), 1);
```

Comment:
maybe this would be to_u32 then?

Reply:
I think I chose usize because that's what's used when it's sent over the C API callback which is what the numeric value is for.